core/src/main/scala/org/apache/spark/rdd/ParallelCollectionRDD.scala

```diff
@@ -127,18 +127,15 @@ private object ParallelCollectionRDD {
       })
     }
     seq match {
-      case r: Range.Inclusive => {
-        val sign = if (r.step < 0) {
-          -1
-        } else {
-          1
-        }
-        slice(new Range(
-          r.start, r.end + sign, r.step).asInstanceOf[Seq[T]], numSlices)
-      }
       case r: Range => {
-        positions(r.length, numSlices).map({
-          case (start, end) =>
+        // 1 to Int.MaxValue and (-2 to Int.MinValue by -1) can trigger exclusive range int overflow
+        val needsInclusiveRange = r.isInclusive && (r.end == Int.MaxValue || r.end == Int.MinValue)
+        positions(r.length, numSlices).zipWithIndex.map({
+          // we need an inclusive range to avoid int overflow and setting the last range to be
+          // inclusive is suffice 
+          case ((start, end), index) if needsInclusiveRange && index == numSlices - 1 =>
+            new Range.Inclusive(r.start + start * r.step, r.end, r.step)
+          case ((start, end), _) =>
             new Range(r.start + start * r.step, r.start + end * r.step, r.step)
```

Comment:
I think as long as we don't change the behavior it's preferrable to rewrite it in a readable manner. Here it's pretty clear to me that if the range is inclusive we should include the last element in the last slice, regardless of whether the range ends in a special value like `Int.MaxValue`.